### PR TITLE
Analog Buffers, Analog Converters and Linker Enhancement

### PR DESCRIPTION
**Pull request by ragv**
#### Todo
- [x] Linker Enhancement using callback to bind function with connection.
- [x] Linker Modification to link at Bus level itself ( Better efficiency )
- [x] Analog to Digital Converter [ SAR - Based  ADC / IEEE 754 Converters]
- [x] Digital to Analog Converter
- [x] Attenuable Analog Buffer Block
- [x] Tests
#### Summary of Changes

Most of the blocks implemented in this PR reflects the functionality and usability of the linker module.
##### 1. Analog to Digital Converter

IEEE64 IEEE32 Generic Signed / Unsigned 16, 8 , 4 bit outputs
'Links' the passed input with analog connector.
Uses AnalogBuffer modules in input and output.
##### 2. Digital to Analog Converter

```
Links passed Bus to IEEE64 IEEE32 Generic Signed / Unsigned 16, 8 , 4 bit inputs
Links an analog Bus to output.
Uses AnalogBuffer modules in input and output.
```
##### 3. Linker Enhancement

The following  enhancements are made to linker module:
- The linker will now bind DiGraph edges ( connection links ) with a method that will be invoked before updation of  the link.
- AutoUpdater is separated from the `connections_updater` which runs as thread. AutoUpdater is a container for the Connections DiGraph and static methods to update the same.
###### SAMPLE  CODE:

```
.
.
def trigger(self):
    augmented_inputs = [ i*0.2 for i in self.inputs ]
    outputs.set_logic_all(augmented_inputs)
.
.
.
def set_inputs(self, inputs):
    AutoUpdater.remove_link(self.inputs)     # Removes all links to / from self.inputs
    AutoUpdater.add_link(inputs, self.inputs, bind_to = ModuleName.trigger, params = [ self ] )
    # This essentially makes the connection_updater_thread call the self.trigger() before updation.
```
##### 4. Analog Buffer Bank Module [ `buffers.py` ]

```
A multi-input externally enabled attenuation ( in voltage db ) capable Buffer is implemented.
This module Helps to link external inputs with input logic.
Overall this is how Modules are implemented using the Busses, Buffer, Linker and trigger.
```

```
                                                       MODULE
                           __________________________________________________________________
ExternalInputs-----------> |--- [ Buffer ] --->[ Inner Transfer Function ]--->[Buffer ] --- |------------>ExternalOutputs
         |          |      |_______|_______________________|_____________________|__________|       |         |
         |          |              |                       |                     |                  |         |
         V          V              V                       V                     V                  V         V
        Bus ------[AU]------------BM----------------trigger block----------------BM---------------[AU]-------Bus
                    |                                   *                                           |
                    |________________callback___________|__________callback_________________________|



*AU - AutoUpdater - linker.py
*BM - AnalogBuffer - buffers.py
```

Sample Code for A2D and D2A Converters [ From `__doc__` string ]

For IEEE 754 Conversions Refer - [Wikipedia](http://en.wikipedia.org/wiki/IEEE_floating_point) and [this](http://babbage.cs.qc.cuny.edu/IEEE-754.old/Decimal.html)
A2D

```
    >>> input_analog   = Bus(Connector(voltage = 6.4))
    >>> input_analog.set_type(analog = True)
    >>> output_digital = Bus(16)
    >>> VREF           = Connector(voltage = 5.0)
    >>> GND            = Connector(voltage = 0)
    >>> a2d_16bit = A2D(input_analog, output_digital, 3, VREF, GND, scale = 0.5)
    >>> time.sleep(0.5) # To allow conversion to take place.
    >>> print output_digital.get_logic_all(as_list = False)

    0b1010001111010111

    >>> input_analog[0].set_voltage(4.2)
    >>> time.sleep(0.5) # To allow conversion to take place.
    >>> print output_digital.get_logic_all(as_list = False)

    0b0110101110000101

    >>> ieee_64bit = Bus(64)
    >>> a2d_IEEE64 = A2D(input_analog, ieee_64bit, 5)
    >>> time.sleep(0.5) # To allow conversion to take place.
    >>> print ieee_64bit.get_logic_all(as_list = False)

    0b0100000000010000110011001100110011001100110011001100110011001101

```

D2A

```

    >>> output_analog   = Bus(Connector(voltage = 6.4))
    >>> output_analog.set_type(analog = True)
    >>> input_digital = Bus(16)
    >>> input_digital.set_logic_all('0110101110000101')
    >>> VREF           = Connector(voltage = 5.0)
    >>> GND            = Connector(voltage = 0)
    >>> d2a_16bit = D2A(input_digital, output_analog, 3, VREF, GND, scale = 2)
    >>> time.sleep(0.1) # To allow conversion to take place.
    >>> print output_analog[0].get_voltage()

    4.19998168945

    >>> ieee_64bit = Bus(64)
    >>> ieee_64bit.set_logic_all('0b0100000000011001100110011001100110011001100110011001100110011010')
    >>> ieee_packed = Bus(1)
    >>> ieee_packed.set_type(analog = True)
    >>> d2a_ieee64 = D2A(ieee_64bit, ieee_packed, 5)
    >>> time.sleep(0.1)
    >>> print  ieee_packed[0].get_voltage()

```

### CODE REVIEW

File: BinPy/analog/buffers.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        Negative value indicates gain.
+        """
+        with AutoUpdater._lock:
+            if type(attenuation) not in [float, int]:
+                raise Exception(
+                    "ERROR: Attenuation can be a float value only.")
+            self.attenuation = float(attenuation)
+            self.trigger()
+
+    def get_attenuation(self):
```
Comment (debugger22) [inline BinPy/analog/buffers.py]: This as a property would be better.

File: BinPy/analog/buffers.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                AutoUpdater.add_link(
+                    [enable],
+                    self.enable,
+                    bind_to=AnalogBuffer.trigger,
+                    params=[self])
+            else:
+                raise Exception(
+                    "ERROR: Invalid Enable input. Enable must be a 1-bit Bus or a Connector.")
+
+    def is_enabled(self):
```
Comment (debugger22) [inline BinPy/analog/buffers.py]: as property.

File: BinPy/analog/buffers.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        """
+        Return true if the buffer block is enabled
+        """
+        with AutoUpdater._lock:
+            return (self.enable[0].get_logic() == 1)
+
+    def is_disabled(self):
```
Comment (debugger22) [inline BinPy/analog/buffers.py]: as property.

File: BinPy/analog/components.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    """
+
+    def __init__(self, params):
+        self.params = {'+': Connector(0), '-': Connector(0), 'i': 0, 'r':
+                       0}
+        for i in params:
+            self.params[i] = params[i]
+
+    def setResistance(self, value):
```
Comment (debugger22) [inline BinPy/analog/components.py:35]: You should follow new naming conventions in this file.

File: BinPy/connectors/connector.py [hunk 5]
```diff
@@ -75,9 +82,12 @@ def untap(self, element, mode):
                 mode)
 
     def set_logic(self, val):
+
         with self._lock:
+            if not self._enable:
+                return
 
-            if type(val) in [int, None, bool]:
+            if type(val) in [int, type(None), bool]:
```
Comment (debugger22) [inline BinPy/connectors/connector.py]: This is interesting. Why'd you do that?

File: BinPy/connectors/connector.py [hunk 6]
```diff
@@ -163,6 +179,20 @@ def __bool__(self):
...
 
+    def enable(self):
+        with self._lock:
+            self._enable = True
+
+    def disable(self):
+        with self._lock:
+            self._enable = False
+
+    def is_enabled(self):
```
Comment (debugger22) [inline BinPy/connectors/connector.py]: property.

File: BinPy/connectors/connector.py [hunk 7]
```diff
@@ -163,6 +179,20 @@ def __bool__(self):
...
+        return self._enable
+
+    def is_disabled(self):
```
Comment (debugger22) [inline BinPy/connectors/connector.py]: property.

**Comment by debugger22** [comment]
@rvraghav93 I haven't tested it yet. Will do so as soon as I get time.

File: BinPy/connectors/connector.py (hunk 5 above)
Comment (ragv) [inline BinPy/connectors/connector.py]: `type(val)` if `val` is `None` is `NoneType` ... which is not equal to `None` :)

Comment (debugger22) [inline BinPy/connectors/connector.py]: Oh I didn't know that.

File: BinPy/analog/converters.py [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+                    - 5 : 64 bit output [ IEEE-754 Floation Point Format ]
+    refp            - Positive reference.
+    refn            - Negative reference.
+    scale           - scale factor to be multiplied with the voltage before conversion.
+    valid           - 1-Bit digital Bus Indicating the end of conversion.
+
+    USAGE
+    =====
+
+        >>> input_analog   = Bus(Connector(voltage = 6.4))
```
Comment (debugger22) [inline BinPy/analog/converters.py]: Indentation.

File: BinPy/analog/converters.py [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+
+        self.set_inputs(analog_input)
+        self.set_outputs(digital_output)
+
+        # Update the values.
+        self.trigger()
+
+    @property
+    def possible_states(self):
+        return float(2 ** (self.outputs.width))
```
Comment (debugger22) [inline BinPy/analog/converters.py]: Why is it necessary to be a floating point?

Comment (ragv) [inline BinPy/analog/converters.py]: Ah ! not needed ! Ill correct that ... !

File: BinPy/analog/converters.py [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+
+        self.set_outputs(analog_output)
+        self.set_inputs(digital_inputs)
+
+        # Update the values.
+        self.trigger()
+
+    @property
+    def possible_states(self):
+        return float(2 ** (self.inputs.width))
```
Comment (debugger22) [inline BinPy/analog/converters.py]: Here also.

File: BinPy/tests/test_analog_converters_buffers.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            start = time.time()
+            while (not bool(test_d2a.valid[0])):
+                now = time.time() - start
+                # If conversion takes more than 1 second Assert False
+                if now > 1:
+                    assert False
+
+            with AutoUpdater._lock:
+                # print ( typ, test_a2d.resolution )
+                # print ( float(analog_ip[0]) , digital_ip.get_logic_all(as_list = False), float(analog_op[0]) )
```
Comment (debugger22) [inline BinPy/tests/test_analog_converters_buffers.py]: You should remove scaffolding.

File: BinPy/tools/clock.py [hunk 12]
```diff
@@ -8,28 +8,43 @@ class Clock(threading.Thread):
...
 
         If time_period and frequency both have been provided, then time_period
         will override frequency
         If nothing is provided, then it will set time_period = 1s by default
-    Methods :   start(), getState(), setState(value), getName(), getTimePeriod(), kill()
+
+    METHODS
+    =======
+
+    start(), getState(), setState(value), getName(), getTimePeriod(), kill()
```
Comment (debugger22) [inline BinPy/tools/clock.py]: You should also update the docs as method names have changed.

File: BinPy/analog/buffers.py (hunk 1 above)
Comment (ragv) [inline BinPy/analog/buffers.py]: Done

File: BinPy/analog/buffers.py (hunk 2 above)
Comment (ragv) [inline BinPy/analog/buffers.py]: Done

File: BinPy/analog/buffers.py (hunk 3 above)
Comment (ragv) [inline BinPy/analog/buffers.py]: Done

File: BinPy/analog/components.py (hunk 4 above)
Comment (ragv) [inline BinPy/analog/components.py:35]: I will rewrite this file fully in subsequent PR Which should include a few other components and the base classes.

File: BinPy/tests/test_analog_converters_buffers.py (hunk 11 above)
Comment (ragv) [inline BinPy/tests/test_analog_converters_buffers.py]: Done

File: BinPy/tools/clock.py (hunk 12 above)
Comment (ragv) [inline BinPy/tools/clock.py]: Done

File: BinPy/tools/clock.py [hunk 13]
```diff
@@ -8,28 +8,43 @@ class Clock(threading.Thread):
...
+    ====
+
+    Once you are done with the clock, use myClock.kill() to kill the clock.
+    Running too many clocks will unnecessarily overload the CPU.
+
+    All operations are thread safe and synchronized between inter / intra thread calls.
+
     """
 
+    _lock = threading.RLock()
```
Comment (ragv) [inline BinPy/tools/clock.py]: @debugger22 Will this cause all the clock instances to share  a common lock ?

File: BinPy/analog/converters.py (hunk 8 above)
Comment (ragv) [inline BinPy/analog/converters.py]: Done

File: BinPy/analog/converters.py (hunk 9 above)
Comment (ragv) [inline BinPy/analog/converters.py]: Done

File: BinPy/analog/converters.py (hunk 10 above)
Comment (ragv) [inline BinPy/analog/converters.py]: Done

File: BinPy/connectors/connector.py (hunk 7 above)
Comment (ragv) [inline BinPy/connectors/connector.py]: Done

File: BinPy/connectors/connector.py (hunk 6 above)
Comment (ragv) [inline BinPy/connectors/connector.py]: Done

File: BinPy/tools/clock.py (hunk 13 above)
Comment (ragv) [inline BinPy/tools/clock.py]: I am removing all locks for now ... If there is a need we can re implement it ...

**Comment by ghost** [comment]
@debugger22 @salil93 @sarwarc Pl review !

**Comment by axiom24** [comment]
Seems good to me. I'm merging this.

**Pull request merged**